### ISSUE TITLE
Add router role with automatic proxying of the crud API via public functions

### ISSUE THREAD

**Issue by akudiyar** [tarantool/crud#26]
Suppose we have a sharded space with customer profiles and want to access them by IPROTO (for example, from a connector). In this case, we have to write a lot of boilerplate code like that:

```lua
...
local function profile_storage_select(conditions)
    return crud.select('profiles', conditions)
end

local function profile_storage_get(id)
    return crud.get('profiles', id)
end

local function profile_storage_insert(tuple)
    return crud.insert('profiles', tuple)
end
...
local function init(opts)
    if opts.is_master then
        box.schema.func.create('profile_storage_insert', {if_not_exists = true})
        box.schema.func.create('profile_storage_get', {if_not_exists = true})
        box.schema.func.create('profile_storage_select', {if_not_exists = true})
        ....
    end

    rawset(_G, 'profile_storage_insert', profile_storage_insert)
    rawset(_G, 'profile_storage_get', profile_storage_get)
    rawset(_G, 'profile_storage_select', profile_storage_select)
    ...

    return true
end

return {
    role_name = 'app.roles.api_storage',
    init = init,
    utils = {
        profile_storage_get = profile_storage_get,
        profile_storage_insert = profile_storage_insert,
        profile_storage_select = profile_storage_select,
        ...
    },
    dependencies = {
        ....
    }
}
```

And we need to do this for each space we want to access using CRUD.

I suggest adding a small module, which encapsulates all this boilerplate logic so that all we need to write is something like

```lua
local proxy = require('proxy').new()

local function init(opts)
    if opts.is_master then
        proxy.register('profiles', 'customers')
    ...
    end
    ...
    return true
end

return {
    role_name = 'app.roles.api_storage',
    init = init,
    utils = table.merge(proxy.api(), {
        ...
    }),
    dependencies = {
        ....
    }
}
```

This module can be automatically enabled in a special "proxy" role when ddl is installed.

**Comment by knazarov** [comment]
Having to do all the boilerplate you list is a problem. But I see a better solution from what you propose.

I think client libraries should call "crud.*" directly, without any wrappers. For this to work, the crud module should register itself in a global variable with the same name as the module name when first loaded. Then it will be possible to call "crud.*" directly.

**Comment by akudiyar** [comment]
@knazarov I will accept either of these solutions, but exposing the whole module seems to be less safe than a list of proxy functions.

**Comment by knazarov** [comment]
@akudiyar exposing the `crud` functions as a global variable is not enough for the client to be able to call them. The client should still have permissions to call those functions. And those permissions must either be explicitly granted, or the user should have execute permission on the `universe`.

If you need granular access control, this is not something I will expose through the data access API. Instead, I propose to add full RBAC to the enterprise edition of this module. The data access API itself should not care whether you have RBAC or not. It should be an implementation detail.

Of course, users are free to roll out their own access control if they want to. For this, they would need to wrap functions as you propose.

**Comment by Totktonada** [comment]
@akudiyar Is there anything else on the topic you want to be implemented aside of PR #58 (released in crud 0.3.0)?

**Comment by akudiyar** [comment]
This topic may be closed, all changes are implemented.
